test/test_layernorm.f90

```diff
@@ -0,0 +1,193 @@
+    end if
+    gradient_flat = reshape(layernorm_instance % gradient, shape(gradient_flat))
+    if (.not. allclose(gradient_flat, expected_gradient_flat)) then
+      ok = .false.
+      write(stderr, '(a)') 'backward returned incorrect gradient values.. failed'
+    end if
+
+    if (.not. allclose(layernorm_instance % d_gamma, expected_d_gamma)) then
+      ok = .false.
+      write(stderr, '(a)') 'backward returned incorrect d_gamma values.. failed'
+    end if
+    if (.not. allclose(layernorm_instance % d_beta, expected_d_beta)) then
+      ok = .false.
+      write(stderr, '(a)') 'backward returned incorrect d_beta values.. failed'
+    end if
+  end subroutine test_layernorm_backward
+
+  subroutine test_layernorm_gradients(input, gradient, ok)
+    real, intent(in out) :: input(:, :)
+    real, intent(in out) :: gradient(:, :)
+    logical, intent(in out) :: ok
+    type(layernorm_layer) :: layernorm_instance
+    type(sgd) :: optim
+
+    real :: parameters(8)
+    real :: expected_parameters(8)
+    real :: updated_output(12)
+    real :: expected_updated_output(12) = [&
+        -0.738849819, 0.881645918, -1.03555739,&
+        1.66299772, -1.02966857, 0.908487320,&
+        -0.562230229, 1.01311040, 0.984123051,&
+        -0.564699769, -1.13543355, -1.11444426&
+    ]
+
+    layernorm_instance = layernorm_layer()
+    call layernorm_instance % init([3, 4])
+
+    call layernorm_instance % forward(input)
+    call layernorm_instance % backward(input, gradient)
+
+    if (layernorm_instance % get_num_params() /= 8) then
+      ok = .false.
+      write(stderr, '(a)') 'incorrect number of parameters.. failed'
+    end if
+
+    expected_parameters(1: 4) = 1.
+    expected_parameters(5: 8) = 0.
+    parameters = layernorm_instance % get_params()
+    if (.not. all(parameters.eq.expected_parameters)) then
+      ok = .false.
+      write(stderr, '(a)') 'incorrect parameters.. failed'
+    end if
+
+    optim = SGD(learning_rate=0.01)
+    call optim % minimize(parameters, layernorm_instance % get_gradients())
+    call layernorm_instance % set_params(parameters)
+
+    call layernorm_instance % forward(input)
+
+    updated_output = reshape(layernorm_instance % output, [12])
+    if (.not. allclose(updated_output, expected_updated_output)) then
+      ok = .false.
+      write(stderr, '(a)') 'incorrect output after parameters update.. failed'
+    end if
+  end subroutine test_layernorm_gradients
+
+  subroutine test_layernorm_integration(ok)
+    logical, intent(in out) :: ok
+
+    type(network) :: net
+    real :: x(2, 3) = reshape([0.1, 2., 0.3, 4., 0.5, 6.], [2, 3])
+    real :: y(6) = [0.7, 0.2, 0.1, 0.1, 0.01, 0.9]
+    real :: tolerance = 0.1
+    integer :: epoch
+    integer :: epochs = 10000
+
+    net = network([&
+        input(2, 3),&
+        linear2d(3),&
+        layernorm(),&
+        flatten()&
+    ])
+
+    ! Kaiming weights to achieve semblance of convergance
```

Comment:
If we want LayerNorm used as activation function to converge, we need to at least use correct weights initialization. Otherwise, the net will not be able to consistently converge and we'll get https://github.com/modern-fortran/neural-fortran/issues/145 once more